### PR TITLE
[custom_op] use new python custom ops API on prims ops

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124665

Also ads a non-decorator version of `custom_op`.

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @zou3519

### CODE REVIEW

File: torch/_inductor/inductor_prims.py [hunk 1]
```diff
@@ -87,6 +89,7 @@ def eager_force_stride(input_tensor: Tensor, stride) -> Tensor:
         self, indices, values, accumulate
     ),
     doc="Unsafe index_put_ (doesn't issue device asserts)",
+    mutates_args=("self",),
```
Comment (zou3519) [inline torch/_inductor/inductor_prims.py:92]: We can infer this from the schema -- feel free to just parse the schema (torch._C.parse_schema). If you're a pytorch developer writing code that's inside pytorch you probably know about alias annotations

File: torch/_dynamo/create_parameter_op.py [hunk 2]
```diff
@@ -19,6 +19,7 @@
     ),
     impl_aten=lambda self, placeholder: placeholder.set_(self),
     doc=doc,
+    returns_alias=True,
```
Comment (zou3519) [inline torch/_dynamo/create_parameter_op.py:22]: This one's suspicious: why does the op return an alias if it's not mentioned in the schema?

File: torch/_prims/__init__.py [hunk 3]
```diff
@@ -302,9 +302,24 @@ def _backend_select_impl(*args, **kwargs):
             return _prim_impl(*args, **kwargs)
 
     name = schema.split("(")[0]
-    prim_impl.impl(name, _prim_impl)
-    prim_autograd_impl.impl(name, _autograd_impl)
-    prim_meta_impl.impl(name, meta)
+    schema = schema[len(name) :]
+
+    # if op returns views/aliases of inputs or other returns, we need to use the
+    # old custom ops API
```
Comment (zou3519) [inline torch/_prims/__init__.py:308]: In theory we can just infer this from the schema as well

File: torch/_prims/__init__.py [hunk 4]
```diff
@@ -1968,7 +1990,7 @@ def _reshape_meta(a: TensorLikeType, shape: ShapeType):
 
 
 def _reshape_aten(a: Tensor, shape: ShapeType) -> Tensor:
-    return a.reshape(shape).contiguous().clone()
+    return a.reshape(shape).contiguous()
```
Comment (zou3519) [inline torch/_prims/__init__.py:1993]: reshape is a bit of a special op. We don't want to constrain it to the custom_op API

**Review by zou3519 — COMMENTED** [review]
Out-of-place custom op (torch._library.custom_ops.py changes) looks good to me. For the make_prims helper functions, we can probably automatically figure out which API we should go through so that users don't need to specify us redundant information

File: torch/_prims/__init__.py [hunk 5]
```diff
@@ -2363,7 +2388,7 @@ def _make_reduction_prim(name: str, impl_aten, doc):
 def _make_var_reduction_prim(name: str, impl_aten, doc):
     """Creates a reduction prim."""
     return _make_prim(
-        schema=f"{name}(Tensor inp, int[]? dims, *, float correction, ScalarType? output_dtype=None) -> Tensor",
+        schema=f"{name}(Tensor inp, int[]? dims, float? correction=1, *, ScalarType? output_dtype=None) -> Tensor",
```
[the remark this answers is not in the document]
Comment (williamwen42) [reply torch/_prims/__init__.py:2391]: This was done in order to get `prims.var` working with the custom ops API. `torch.var` has the kwarg-only argument `correction` that was causing issues.

File: torch/_prims/__init__.py (hunk 4 above)
Comment (williamwen42) [reply torch/_prims/__init__.py:1993]: Do you mean to add back the `.clone()`? Or to use the old custom op API?

Comment (zou3519) [reply torch/_prims/__init__.py:1993]: Oh I see, I misunderstood. We should add the clone back

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**